### ISSUE TITLE
Coordinate systems

### ISSUE THREAD

**Issue by jakubcerveny** [graphdeco-inria/gaussian-splatting#100]
Hi, thanks for making this amazing work available!

I'm trying to convert my own data to the NeRF format so I can process it with `train.py`, but I'm having some trouble with the coordinate systems.

Do the NeRF import functions follow the conventions described here: https://docs.nerf.studio/en/latest/quickstart/data_conventions.html ?

I.e., is the `transform_matrix` expected to contain columns with camera X (right), Y (up), Z (back) axes, and the camera position? Inside your code I see it is inverted to become the world-to-camera matrix, but what is the meaning of the transpose and the sign flips in `readCamerasFromTransforms:191-193`? 

I have several cameras that share the same position, i.e., they have the same last column in `transform_matrix`, but in the SIBR viewer they no longer keep the same position :-/

Also, in what coordinate system is the initial point cloud expected? I suspect internally the COLMAP system is used for the world coordinates (Y down, Z forward), is that correct?

Thank you,
Jakub

**Comment by BennetLeff** [comment]
Some clarity on this would be great!

**Comment by grgkopanas** [comment]
Hi,

Let me try to clarify as much as I can the camera systems we use:
 * In the python code and the cuda gaussian rasterizer we use the colmap camera system: Y-down Z-forward. This can be seen [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/main/scene/dataset_readers.py#L82-L83), where we read the colmap files and we do no transformations. While [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/main/scene/dataset_readers.py#L191) you see how we trasnform the "nerf-synthetic" camera model to the colmap one.

 * The sIBR viewer has a different camera model and we transform the cameras from sIBR to COLMAP camera model here: https://github.com/graphdeco-inria/gaussian-splatting/blob/main/gaussian_renderer/network_gui.py#L75C1-L76C67

> Also, in what coordinate system is the initial point cloud expected? I suspect internally the COLMAP system is used for the world coordinates (Y down, Z forward), is that correct?

In this case I am not sure what a "point-cloud coordinate system" means I assume that the above probably clarify what you need. 

Feel free to explain more in case things dont work out for you, I know that some parts of our code regarding camera systems are very obscure and do a lot of unnecessary operations, it's on my TODO list to clean it up.

**Comment by BennetLeff** [comment]
Great, this is what I ascertained after digging into it for a long time but want confirmation because my renders with custom data aren't turning out right.

**Comment by grgkopanas** [comment]
Just a head's up but you probably already realized for legacy reasons we feed the world2camera and the full_transformation matrix transposed to the cuda rasterizer. This has to do with row-major and column-major assumptions between our cuda code and python code.

**Comment by grgkopanas** [comment]
hmm, the intention where I guess would be to negate Y and Z axes.[ In line 191](https://github.com/graphdeco-inria/gaussian-splatting/blob/main/scene/dataset_readers.py#L191) we negate all 3 axeses but then [we negate back](https://github.com/graphdeco-inria/gaussian-splatting/blob/main/scene/dataset_readers.py#L192) the axis X .

This could possible have some hidden errors, if you find one and you can actually fix it, feel free to make a pull request and as long as the nerf synthetic datasets still works we will merge it.

-gk

**Comment by cgallik** [comment]
Any way you could use the initial transforms.json in nerfstudio format (using iphone's arkit in something like nerfcapture or kiri engine) to initialize camera poses in colmap?

https://colmap.github.io/faq.html#reconstruct-sparse-dense-model-from-known-camera-poses

been trying to make this work but to no avail.

**Comment by grgkopanas** [comment]
I am not very familiar with the details of all the software that you
mentioned so I don't feel very comfortable advising on that.

-gk

**Comment by BennetLeff** [comment]
> Any way you could use the initial transforms.json in nerfstudio format (using iphone's arkit in something like nerfcapture or kiri engine) to initialize camera poses in colmap?
> 
> https://colmap.github.io/faq.html#reconstruct-sparse-dense-model-from-known-camera-poses
> 
> been trying to make this work but to no avail.

Have you read the nerfstudio documentation on coordinate systems? It fills in some of the info to get started digging into this. The short answer is that yes, the `transforms.json` file contains the necessary information but actually accomplishing that has proven non-trivial.

**Comment by coreqode** [comment]
Just a small doubt.  Why are we transposing the rotation matrix [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40) ?

**Comment by grgkopanas** [comment]
> Just a small doubt. Why are we transposing the rotation matrix [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40) ? Is it because we are converting to Column-major convention?

I have to clean this up when I find time. No this transpose is to cancel out another transpose that never should have been there.

To accommodate column-major convention we transpose [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/scene/cameras.py#L54)

**Comment by Bin-ze** [comment]
> Any way you could use the initial transforms.json in nerfstudio format (using iphone's arkit in something like nerfcapture or kiri engine) to initialize camera poses in colmap?
> 
> https://colmap.github.io/faq.html#reconstruct-sparse-dense-model-from-known-camera-poses
> 
> been trying to make this work but to no avail.

I have implemented it, the details can be viewed：
https://github.com/Bin-ze/3d_gaussian_magic_change/blob/master/scene/dataset_readers.py

**Comment by cgallik** [comment]
> > Any way you could use the initial transforms.json in nerfstudio format (using iphone's arkit in something like nerfcapture or kiri engine) to initialize camera poses in colmap?
> > https://colmap.github.io/faq.html#reconstruct-sparse-dense-model-from-known-camera-poses
> > been trying to make this work but to no avail.
> 
> I have implemented it, the details can be viewed： https://github.com/Bin-ze/3d_gaussian_magic_change/blob/master/scene/dataset_readers.py

First off, thank you!
secondly, I only think you are halfway there. You read the data from the transfroms.json file for GS, but i think my idea was to use use it to help initialize colmap's SfM pipeline. The link I provided explains how to do this. 

My reasoning for doing this is, if I have the initial camera positions from my iphone using arkit enabled apps. It is my understanding that this could help colmap in the sparse cloud reconstruction, and be refined in the bundle adjustment stage. Then instead of initializing GS from random points and using camera positions from the transforms.json file, you can have the sparse cloud that is in the standard pipeline, and be you'll be a little more certain that colmap will work!

Ideally, I want to set up a sort of "rig" with my iphone and a more professional camera side-by-side. 

I also see you are working on bringing in pointcloud from polycam, this would be another interesting way of getting points and camera positions!

**Comment by Snosixtyboo** [comment]
Should be resolved with acceptance of @jakubcerveny 's pull request

**Comment by coreqode** [comment]
I realized that `getProjectionMatrix` and `focal2fov` are only valid when `cx` and `cy` are center of the image pixels. If we get unbalanced camera centers from the calibration, we might get different `fovs` and `view transform matrix`?  Is this a valid assumption?

**Comment by pknmax** [comment]
@jakubcerveny 

How to interpret view direction and view dependent colors computation, in colmap camera system:

camera:
def __init__(self, uid, R, T, FoVx, FoVy, image_width, image_height):
        self.uid = uid
        self.R = R
        self.T = T
        self.FoVx = FoVx
        self.FoVy = FoVy
        self.image_width = image_width
        self.image_height = image_height
        self.projection_matrix = getProjectionMatrix(znear=0.01, zfar=100.0, fovX=FoVx, fovY=FoVy).transpose(0,1).cuda()
        self.world_view_transform = torch.tensor(getWorld2View2(R, T, np.array([0,0,0]), 1.0)).transpose(0, 1).cuda()
        self.full_proj_transform = (self.world_view_transform.unsqueeze(0).bmm(self.projection_matrix.unsqueeze(0))).squeeze(0)
        self.camera_center = self.world_view_transform.inverse()[3, :3]
        
view dependent color  computation:
        shs_view = pc.get_features.transpose(1, 2).view(-1, 3, (pc.max_sh_degree+1)**2)
        dir_pp = (pc.get_xyz - viewpoint_camera.camera_center.repeat(pc.get_features.shape[0], 1))
        dir_pp_normalized = dir_pp/dir_pp.norm(dim=1, keepdim=True)
        sh2rgb = eval_sh(pc.active_sh_degree, shs_view, dir_pp_normalized)
        colors_precomp = torch.clamp_min(sh2rgb + 0.5, 0.0)
        colors_list.append(colors_precomp[k].detach().cpu().numpy())
        
 rotation matrix to eular angles (is it correct?):
    Rx = np.arctan2(-R[1, 2], R[2, 2])
    Ry = np.arcsin(R[0, 2])
    Rz = np.arctan2(-R[1, 0], R[0, 0])
     
i want to tag camera pose and respective view dependent color of gaussian, how i can identify this relationship?
I find Rx, Ry and Rz very nonintuitive,

suppose I facing the object, and moving the camera up then down, then moved to the right from current position and then moving the camera up then down,
then how should  Rx, Ry and Rz change?

**Comment by hughkhu** [comment]
> > Just a small doubt. Why are we transposing the rotation matrix [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40) ? Is it because we are converting to Column-major convention?
> 
> I have to clean this up when I find time. No this transpose is to cancel out another transpose that never should have been there.
> 
> To accommodate column-major convention we transpose [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/scene/cameras.py#L54)

I find [dataset_readers.py#L191](https://github.com/graphdeco-inria/gaussian-splatting/blob/2eee0e26d2d5fd00ec462df47752223952f6bf4e/scene/dataset_readers.py#L191). Is it the reason that we transpose R [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40)

**Comment by rookie-oops** [comment]
> > > Just a small doubt. Why are we transposing the rotation matrix [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40) ? Is it because we are converting to Column-major convention?
> > 
> > 
> > I have to clean this up when I find time. No this transpose is to cancel out another transpose that never should have been there.
> > To accommodate column-major convention we transpose [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/scene/cameras.py#L54)
> 
> I find [dataset_readers.py#L197](https://github.com/graphdeco-inria/gaussian-splatting/blob/2eee0e26d2d5fd00ec462df47752223952f6bf4e/scene/dataset_readers.py#L197). Is it the reason that we transpose R [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40) ?
I'm also puzzled by this question. Do you have any idea now?

**Comment by AsherJingkongChen** [comment]
> > > > Just a small doubt. Why are we transposing the rotation matrix [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40) ? Is it because we are converting to Column-major convention?
> > > 
> > > 
> > > I have to clean this up when I find time. No this transpose is to cancel out another transpose that never should have been there.
> > > To accommodate column-major convention we transpose [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/scene/cameras.py#L54)
> > 
> > 
> > I find [dataset_readers.py#L197](https://github.com/graphdeco-inria/gaussian-splatting/blob/2eee0e26d2d5fd00ec462df47752223952f6bf4e/scene/dataset_readers.py#L197). Is it the reason that we transpose R [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/utils/graphics_utils.py#L40) ?
> > I'm also puzzled by this question. Do you have any idea now?

Yes, obviously.

**Comment by Yeojin-Lee-00** [comment]
Hello,

I am using the ShapeNet SRN dataset's cars scene for training with random initialization in Blender format. However, I encountered an issue during training, and after searching the issues, I found your post and would like to ask for help.

Since the ShapeNet SRN dataset follows the OpenCV coordinate system, I converted the poses of the dataset to the OpenGL format (by flipping the Y and Z axes) and created the transform_train.json and transform_test.json files for training. 

After 30,000 iterations, the point cloud results look like the image I have attached below, but the training does not seem to progress at all.
![image](https://github.com/user-attachments/assets/c92e2b85-f866-4ed2-974c-48684aeef30d)

I printed the values of transform_matrix, and the camera positions are as follows:
![image](https://github.com/user-attachments/assets/af9d98c9-e2de-425d-bd3c-0d770bc39249)

Have you had any success training with random initialization on a custom dataset? Is random initialization for the Blender dataset only effective for the NeRF synthetic dataset?

Thank you.

**Comment by BW-Zhang05** [comment]
> Hello,
> 
> I am using the ShapeNet SRN dataset's cars scene for training with random initialization in Blender format. However, I encountered an issue during training, and after searching the issues, I found your post and would like to ask for help.
> 
> Since the ShapeNet SRN dataset follows the OpenCV coordinate system, I converted the poses of the dataset to the OpenGL format (by flipping the Y and Z axes) and created the transform_train.json and transform_test.json files for training.
> 
> After 30,000 iterations, the point cloud results look like the image I have attached below, but the training does not seem to progress at all. ![image](https://github.com/user-attachments/assets/c92e2b85-f866-4ed2-974c-48684aeef30d)
> 
> I printed the values of transform_matrix, and the camera positions are as follows: ![image](https://github.com/user-attachments/assets/af9d98c9-e2de-425d-bd3c-0d770bc39249)
> 
> Have you had any success training with random initialization on a custom dataset? Is random initialization for the Blender dataset only effective for the NeRF synthetic dataset?
> 
> Thank you.

3dgs need special viewer , meshlab is not available, please use supersplat
